WebAuthn userHandle must be null, not empty string

pascoej opened the pull request:
#### 04d4979c9e1d766b1ad11eac36f7593ee8bb268d
<pre>
WebAuthn userHandle must be null, not empty string
<a href="https://rdar.apple.com/92305724">rdar://92305724</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=239737">https://bugs.webkit.org/show_bug.cgi?id=239737</a>

Reviewed by Brent Fulgham.

Whenever we started calling out to AuthenticationServicesAgent, we started passing no user handle
as an empty ArrayBuffer instead of null. This isn&apos;t allowed by the spec, if there&apos;s no user handle
we should return null. This change does that by treating an empty user handle as null.

This is already checked for in LayoutTests, however, they don&apos;t exercise the
serialization back and forth from the daemon.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::toArrayBuffer):

Canonical link: <a href="https://commits.webkit.org/275669@main">https://commits.webkit.org/275669@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e13e58ff3fe7093bf434d019b6d280e02ea4a2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42488 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44883 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45089 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38605 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18859 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43062 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18421 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/36572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16119 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/46579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/38655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37945 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41848 "layout-tests (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17304 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/14234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18923 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9498 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18568 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -96,8 +96,10 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
+static inline RefPtr<ArrayBuffer> toArrayBuffer(NSData *data)
 {
+    if (!data || !data.length)
+        return nullptr;
```
It's not very obvious how this change fixes the issue you describe in the ChangeLog, because this is such a generically named method. It might be slightly clearer if the location where `toArrayBuffer` was called did the null check, but I think this is fine (especially if it's used in multiple places).

cdumez replied:
How come there is no test?

pascoej committed:
WebAuthn userHandle must be null, not empty string
rdar://92305724
https://bugs.webkit.org/show_bug.cgi?id=239737

Reviewed by Brent Fulgham.

Whenever we started calling out to AuthenticationServicesAgent, we started passing no user handle
as an empty ArrayBuffer instead of null. This isn't allowed by the spec, if there's no user handle
we should return null. This change does that by treating an empty user handle as null.

This is already checked for in LayoutTests, however, they don't exercise the
serialization back and forth from the daemon.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::toArrayBuffer):

Canonical link: https://commits.webkit.org/275669@main
